booster/library/Booster/SMT/Interface.hs

```diff
@@ -178,6 +192,23 @@ finaliseSolver ctxt = do
     Log.logMessage ("Closing SMT solver" :: Text)
     destroyContext ctxt
 
+failBecauseUnknown ::
+    forall io.
+    Log.LoggerMIO io =>
+    Set Predicate ->
+    ExceptT SMTError (SMT io) Bool
+failBecauseUnknown psToCheck =
+    smtRun GetReasonUnknown >>= \case
+        Unknown reason -> do
+            Log.withContext Log.CtxAbort $
+                Log.logMessage $
+                    "Returned Unknown. Reason: " <> fromMaybe "UNKNOWN" reason
+            throwE $ SMTSolverUnknown reason mempty psToCheck
+        other -> do
+            let msg = "Unexpected result while calling ':reason-unknown': " <> show other
+            Log.withContext Log.CtxAbort $ Log.logMessage $ Text.pack msg
+            throwSMT' msg
+
```

Comment:
I previously made it so that we call `GetReasonUnknown` whenever we get an `Unknown` result from the solver. Hence we should never need to call `GetReasonUnknown` explicitly.

Reply:
Sorry should have made this explicit in the type... I think we should change `Response` to:

```haskell
data Response' reason
    = Success -- for command_
    | Sat
    | Unsat
    | Unknown reason
    | Values [(SExpr, Value)]
    | Error BS.ByteString
    deriving stock (Eq, Ord, Show)

type Response = Response' Text
type ResponseMay = Response' (Maybe Text)
```

s.t.

```haskell
runCmd :: forall cmd io. (SMTEncode cmd, LoggerMIO io) => cmd -> SMT io Response
runCmd c = runCmdMaybe c >>= \case
    Unknown Nothing -> pure $ Unknown"Could not determine reason when running GetReasonUnknown"
    Unknown (Just reason) -> pure $ Unknown reason
    Success -> pure Success
    Sat -> pure Sat
    Unsat -> pure Unsat
    Values vs -> pure $ Values vs
    Error e -> pure $ Error e

    
    where
        runCmdMaybe :: forall cmd'. SMTEncode cmd' => cmd' -> SMT io ResponseMay
        runCmdMaybe cmd = do
            let cmdBS = encode cmd
            ctxt <- SMT get
            case ctxt.mbSolver of
                Nothing -> pure $ Unknown (Just "server started without SMT solver")
                Just solverRef -> do
                    whenJust ctxt.mbTranscriptHandle $ \h -> do
                        whenJust (comment cmd) $ \c ->
                            liftIO (BS.hPutBuilder h c)
                        liftIO (BS.hPutBuilder h $ cmdBS <> "\n")
                    output <- (liftIO $ readIORef solverRef) >>= \solver -> run_ cmd solver cmdBS
                    let result = readResponse output
                    whenJust ctxt.mbTranscriptHandle $
                        liftIO . flip BS.hPutStrLn (BS.pack $ "; " <> show output <> ", parsed as " <> show result <> "\n")
                    case result of
                        Error{} -> do
                            logMessage $
                                "SMT solver reports: " <> pack (show result)
                            pure result
                        Unknown Nothing ->
                            runCmdMaybe GetReasonUnknown >>= \case
                                unknownWithReason@(Unknown (Just _)) -> pure unknownWithReason
                                _ -> pure result
                        _ -> pure result
```